llvm/lib/Target/AIE/AIEBaseSubtarget.cpp

```diff
@@ -318,6 +305,56 @@ class RegionEndEdges : public ScheduleDAGMutation {
   };
 };
 
+/// This Mutator is responsible for emitting "fixed" SUnits at the top or bottom
+/// of the region. These special SUnits require a specific cycle and cannot be
+/// placed freely by the scheduler.
+///
+/// Here, these special SUnits get created from Region::top_fixed_instrs() or
+/// Region::bot_fixed_instrs(), and dependencies are created between "free" and
+/// "fixed" SUnits.
+class EmitFixedSUnits : public ScheduleDAGMutation {
+public:
+  void apply(ScheduleDAGInstrs *DAG) override {
+    AIEPostRASchedStrategy *Scheduler =
+        static_cast<AIEScheduleDAGMI *>(DAG)->getSchedImpl();
+    auto *TII = static_cast<const AIEBaseInstrInfo *>(DAG->TII);
+    auto *ItinData = DAG->MF.getSubtarget().getInstrItineraryData();
+    const BlockState &BS =
+        Scheduler->getInterBlock().getBlockState(DAG->getBB());
+    const Region &CurRegion = BS.getCurrentRegion();
+
+    // First, create SUnits for all "fixed" instructions
+    unsigned DistToExitSU = 0;
+    for (MachineInstr &MI : reverse(CurRegion.bot_fixed_instrs())) {
+      Scheduler->addFixedSUnit(MI, /*IsTop=*/false, DistToExitSU);
+      ++DistToExitSU;
+    }
+    DAG->makeMaps();
+
+    // Then, create dependencies between "free" and "fixed" instructions
+    auto IsFreeSU = [Scheduler](const SUnit &SU) {
+      return Scheduler->isFreeSU(SU);
+    };
+    ArrayRef<AIE::MachineBundle> BotFixedBundles =
+        CurRegion.getBotFixedBundles();
+    for (SUnit &FreeSU : make_filter_range(DAG->SUnits, IsFreeSU)) {
+      const MachineInstr &MI = *FreeSU.getInstr();
+      MachineInstr *FixedDepMI =
```

Comment:
super nit:

To help readability, we could extend with overloaded cast operators:

```
 struct InstrAndCycle {
   MachineInstr *MI = nullptr;
   int Cycle;
+  operator MachineInstr *() const { return MI; }
+  operator int() const { return Cycle; }
 };
```
Then we can just use:

```
MachineInstr *FixedDepMI = AIE::findEarliestRef(MI, BotFixedBundles, BotFixedBundles.size());
```

And also:

```
Earliest = findEarliestRef(MI, TopBundles, Earliest);
```